### PR TITLE
Support specifying 'latest' in model URI to get the latest version of a model regardless of the stage

### PR DESCRIPTION
**Pull request by lichenran1234**
Signed-off-by: Chenran Li <chenran.li@databricks.com>

## What changes are proposed in this pull request?

Support specifying 'latest' in model URI like `models:/<model_name>/latest` to get the latest version of a model regardless of the stage.

Also fix a bug in `sqlalchemy_store.get_latest_versions()`: when no `stage` argument is specified, it should return the latest versions of all stages. Now it's only returning latest versions for active stages.
* according to the [top-level comment](https://sourcegraph.com/github.com/mlflow/mlflow@027f75436c514bbfb8eb44607bdad07737cd5736/-/blob/mlflow/protos/model_registry.proto?L521) in the proto file, it should return the latest version of all stages when no `stage` argument is specified
* also fixed the contradictory comments in the docstring of the function:
Screen Shot 2021-11-08 at 17 48 52

In #4250, Anil also suggested supporting model URI formats like `models:/<model_name>/latest-n` to get the nth to last model version. But it requires too big of a change (e.g. extending the [`RegisteredModel`](https://sourcegraph.com/github.com/mlflow/mlflow@b6eaafdfce2d586aaa39185a36edaa7ec343a7c5/-/blob/mlflow/entities/model_registry/registered_model.py?L10) class and the corresponding proto message to store not only the latest versions of a model, but also all versions). It may not be worth it to make such a big change for this small "latest-n" feature. So I'm not doing this in this PR.

## How is this patch tested?

unit tests

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Now users can specify 'latest' in model URI like `models:/<model_name>/latest` to get the latest version of a model regardless of the stage. Previously users can only specify `models:/<model_name>/<Stage>` to get the latest version of a model on a specific stage.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [x] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [x] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Review by ankit-db — COMMENTED** [review]
The change looks good to me - just a few questions! I do think it may qualify as a breaking change though, so we should make sure to change the label

File: mlflow/store/artifact/utils/models.py [hunk 1]
```diff
@@ -13,24 +16,34 @@ def is_using_databricks_registry(uri):
 def _improper_model_uri_msg(uri):
     return (
         "Not a proper models:/ URI: %s. " % uri
-        + "Models URIs must be of the form 'models:/<model_name>/<version or stage>'."
+        + "Models URIs must be of the form 'models:/<model_name>/suffix' "
+        + "where suffix is a model version, stage, or string '%s'." % _MODELS_URI_SUFFIX_LATEST
```
[the remark this answers is not in the document]
Comment (lichenran1234) [reply mlflow/store/artifact/utils/models.py]: Done, thanks!

File: mlflow/store/artifact/utils/models.py [hunk 2]
```diff
@@ -46,12 +59,16 @@ def _parse_model_uri(uri):
 
     if parts[1].isdigit():
```
[the remark this answers is not in the document]
Comment (lichenran1234) [reply mlflow/store/artifact/utils/models.py:60]: Done, thanks!

File: mlflow/store/artifact/utils/models.py [hunk 3]
```diff
@@ -13,24 +16,34 @@ def is_using_databricks_registry(uri):
...
+        stage_str = "" if stage is None else " and stage '{stage}'".format(stage=stage)
         raise MlflowException(
-            "No versions of model with name '{name}' and "
-            "stage '{stage}' found".format(name=name, stage=stage)
+            "No versions of model with name '{name}'{stage_str} found".format(
+                name=name, stage_str=stage_str
+            )
         )
-    return latest[0].version
+    return max(map(lambda x: int(x.version), latest))
```
[the remark this answers is not in the document]
Comment (ankit-db) [reply mlflow/store/artifact/utils/models.py:37]: Got it

**Review by ankit-db — APPROVED** [review]

**Pull request merged**